File: code/C++/128-Longest Consecutive Sequence.cpp

/*
Longest Consecutive Sequence

Given an unsorted array of integers, find the length of the longest consecutive elements sequence.

For example,
Given [100, 4, 200, 1, 3, 2],
The longest consecutive elements sequence is [1, 2, 3, 4]. Return its length: 4.

Your algorithm should run in O(n) complexity.

典型的用空间换时间，用一个巨大的哈希表来存这些数值，
然后依次寻找连续的最长串

*/
class Solution {
public:
    int longestConsecutive(vector<int> &num) {
        unordered_map<int, bool> used;
        for (auto i : num) used[i] = false;
        
        int longest = 0;
          
        for (auto i : num) {
            if(used[i]) continue;
            
            int length=1;
            
            used[i]=true;
            
            for(int j=i+1;used.find(j)!=used.end();j++){
                used[j]=true;
                length++;
            }
            for(int j=i-1;used.find(j)!=used.end();j--){
                used[j]=true;
                length++;
            }
            
            longest=max(longest,length);
            
        }
        
        return longest;
        
        
    }
};

File: code/Python/128-Longest Consecutive Sequence.py
# -*- coding: utf-8 -*

# Given an unsorted array of integers, find the length of the longest consecutive elements sequence.
#
# Your algorithm should run in O(n) complexity.
#
# Example:
#
# Input: [100, 4, 200, 1, 3, 2]
# Output: 4
# Explanation: The longest consecutive elements sequence is [1, 2, 3, 4]. Therefore its length is 4.

class Solution(object):
    def longestConsecutive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length = len(nums)
        if length == 0:
            return 0
        elif length == 1:
            return 1
        nums.sort()
        max_len, len_ = 0, 1
        pre = nums[0]
        for item in nums[1:]:
            if item - 1 == pre:
                len_ += 1
            elif item != pre:
                len_ = 1
            max_len = max(max_len, len_)
            pre = item
        return max_len

if __name__ == '__main__':
    s = Solution()
    nums = [1,2,0,1]
    print(s.longestConsecutive(nums))